src/package-hoister.js

```diff
@@ -649,6 +658,81 @@ export default class PackageHoister {
     }
   }
 
+  markShallowWorkspaceEntries(){
+    const targetWorkspace = path.basename(this.config.cwd);
+    const entries = Array.from(this.tree);
+    
+    //get all workspaces that the target depends on
+    const dependentWorkspaces = entries.filter(([key, info]) => {
+      //don't mark the target workspace or non-workspaces
+      if(key === targetWorkspace || !WorkspaceResolver.isWorkspace(key + '@', this.workspaceLayout)){
+        return false;
+      }
+      //skip the root workspace
+      if(key.includes('workspace-aggregator')){
+        return false;
+      }
+      //skip a workspace if a different version of it is already being installed under the target workspace
+      if(this.tree.has(`${targetWorkspace}#${key}`)){
+        return false;
+      }
+      const searchPath = `/${WS_ROOT_ALIAS}/${targetWorkspace}`;
+      return info.previousPaths.some(p => p.startsWith(searchPath));
+    }).map(e => e[0]);
+
+    entries.forEach(([key, info]) => {
+      const splitPath = key.split('#');
+
+      //mark the workspace and any un-hoisted dependencies it has for shallow installation
+      const isShallowDependency = dependentWorkspaces.some(w => {
+        if(splitPath[0] !== w){ //entry is not related to the workspace
+          return false;
+        }
+        if(!splitPath[1]){ //entry is the workspace
+          return true;
+        }
+        //don't bother marking dev dependencies for shallow installation
+        const pkg = this.tree.get(w).pkg;
+        return !pkg.devDependencies || !(splitPath[1] in pkg.devDependencies);
+      });
+
+      if(isShallowDependency){
+        info.shallowPaths = [''];
+        return;
+      }
+
```

Comment:
Finally, it looks for packages that are normally hoisted but need to be shallowly installed underneath a sibling workspace because the target has its own version, and marks those for shallow installation (but not just directly under node_modules like the other ones)